Generator documentation (#8721)
opsmill/infrahub: docs/docs/topics/generator.mdx

```text
Generator Definition
  │
  ▼
Target Group
  ├── Member A  →  Run A (variables from A)
  ├── Member B  →  Run B (variables from B)
  └── Member C  →  Run C (variables from C)
```

Each run is fully independent — it has its own query variables, its own query results, and its own Generator instance. Runs do not share state.

## Query parameter mapping

The `parameters` field in `.infrahub.yml` controls how Infrahub extracts variables from each target object and passes them to the GraphQL query. This is the mechanism that scopes each run to its target.

### How it works

Given this Generator definition:

```yaml
generator_definitions:
  - name: widget_generator
    file_path: "generators/widget_generator.py"
    targets: widgets
    query: widget_query
    class_name: WidgetGenerator
    parameters:
      name: "name__value"
```

And this GraphQL query:

```graphql
query Widgets($name: String!) {
  TestWidget(name__value: $name) {
    edges {
      node {
        name { value }
        count { value }
      }
    }
  }
}
```

For each member of the `widgets` group, Infrahub:

1. Reads the parameter mapping: `name` → `"name__value"`

> 2. Extracts the value from the target object using the double-underscore path

Reviewer: ```suggestion
2. Extracts the value from the target object using the defined path
```

3. Passes it as a query variable

For example:

| Target object | Extraction path        | Extracted value | Query variable          |
|---------------|------------------------|-----------------|-------------------------|
| `widget1`     | `widget1.name.value`   | `"widget1"`     | `$name = "widget1"`     |
| `widget2`     | `widget2.name.value`   | `"widget2"`     | `$name = "widget2"`     |

Each run's GraphQL query only returns data for its specific target, keeping runs independent.

### Double-underscore notation

The double-underscore (`__`) in parameter values traverses the object hierarchy:

- `name__value`: attribute `name`, property `value`

> - `location__name__value`: relationship `location` (cardinality-one), then attribute `name`, property `value`

Reviewer: Is it really something for generators?

Reviewer: Overall I'm not sure about that relationship use case for generators

Author: It is not specific to generators, you are right. I can't come up with a use case either, but it is possible.

| Layer  | Target group | Members      | Parallel runs |
|--------|--------------|--------------|---------------|
| Fabric | `dc_fabrics` | 1 fabric     | 1             |
| Pod    | `dc_pods`    | 4-8 pods     | 4-8           |
| Rack   | `dc_racks`   | 32+ racks    | 32+           |

The fabric Generator runs once (1 target). It creates pod objects. The pod Generator runs 4-8 times concurrently. Each pod Generator creates rack objects. The rack Generator runs 32+ times concurrently. Total parallelism is multiplicative across layers.

### Guideline: prefer more, smaller targets

A single Generator targeting a group with one member that creates 100 objects runs as one sequential operation. The same work split across 10 targets with 10 objects each runs as 10 concurrent operations — significantly faster on a system with available capacity.

## The relationship between `member_of_groups` and Generator targeting

Objects become Generator targets by being members of the group specified in the Generator definition's `targets` field.

### How objects become targets

You add objects to a target group through the `member_of_groups` relationship, which can be set:

> - **In the UI**: when creating or editing an object, select the group in the "Member of groups" field

Reviewer: It's under "Manage" > "Groups" for edit

Reviewer: For what's it's worth we even have a guide for groups https://docs.infrahub.app/guides/groups

Reviewer: So that section could be "add the object to a group so it becomes a target > here is the guide"

- **In object files**: define group membership in your YAML object definitions
- **Programmatically**: via the SDK or GraphQL mutations

### Dynamic targeting

Adding or removing group members changes what gets targeted on the next Generator run:

- **Add an object to the group**: it becomes a target and gets its own Generator run next time the definition executes
- **Remove an object from the group**: it is no longer targeted (existing generated objects are not automatically cleaned up — see [known limitation #3289](https://github.com/opsmill/infrahub/issues/3289))

### Standard groups vs. Generator groups

These two group types serve different purposes and are often confused:

| Group type            | Purpose                                                                                            | You manage it |
|-----------------------|----------------------------------------------------------------------------------------------------|---------------|
| `CoreStandardGroup`   | Defines which objects are **targeted by** a Generator. Listed in the `targets` field of the Generator definition. | Yes — you create it and control membership |

> | `CoreGeneratorGroup`  | Tracks which objects were **created by** a Generator. Managed automatically by the SDK tracking feature. | No — Infrahub manages this automatically |

Reviewer: ```suggestion
| `CoreGeneratorGroup`  | Tracks which objects were **created by** a Generator instance. Managed automatically by the SDK tracking feature. | No — Infrahub manages this automatically |
```

The target group (`CoreStandardGroup`) is an input to the Generator — "run against these objects." The Generator group (`CoreGeneratorGroup`) is an output — "these objects were created by this Generator."

## Query response modes

The `convert_query_response` flag in `.infrahub.yml` controls how the GraphQL query results are delivered to your `generate()` method. This affects how you access data and what SDK features are available.

### Raw dict mode (default)

When `convert_query_response` is `false` (the default), the `data` parameter passed to `generate()` is the raw GraphQL response dictionary:

```python
class DeviceGenerator(InfrahubGenerator):
    async def generate(self, data: dict) -> None:
        for edge in data["TestDevice"]["edges"]:
            device_name = edge["node"]["name"]["value"]
            device_role = edge["node"]["role"]["value"]
            # ... create objects using self.client
```

> This is the more lightweight mode — no conversion overhead, and you work with plain Python dictionaries. You can also use [Pydantic models](https://docs.pydantic.dev/) to parse the response into typed objects for better IDE support and validation.

Reviewer: Can't recall the status of Pydantic models here, not sure we want to recommend that here

Author: What we mean here is the Pydantic models you can generate from GraphQL responses. I'm not sure if it is documented atm

### SDK object mode

> When `convert_query_response` is `true`, Infrahub converts the GraphQL response into `InfrahubNode` SDK objects. These are available via `self.nodes` (a list) and `self.store` (indexed by node ID). You do **not** use the `data` parameter in this mode:

Reviewer: "`self.store` (indexed by node ID)" > should we keep this?

```python
class DeviceGenerator(InfrahubGenerator):
    async def generate(self, data: dict) -> None:
        for device in self.nodes:
            device_name = device.name.value
            device_role = device.role.value
            # SDK features available: relationships, .save(), .delete()
```

### When to use each

| Use raw dict mode (`false`) when | Use SDK object mode (`true`) when |
|---|---|
| Dict access is sufficient for your use case | You need SDK features like `.save()`, `.delete()`, or relationship traversal on the queried objects |
| You want minimal overhead | You prefer cleaner attribute access (`node.name.value` vs `node["name"]["value"]`) |
| The query returns a flat structure | You want to use `self.store` to look up nodes by ID |

> | You want to use Pydantic models for type-safe parsing | You want the SDK to handle response parsing automatically |

Reviewer: This table isn't clear to me. I guess to me the main differentiator is wether the expected complexity for the generator is high or not. If yes then protocols + convert on. If no then raw dict.

Author: I think it's ok like this

### `execute_in_proposed_change` (default: `true`)

When `true`, the Generator runs during proposed change CI. Results appear in the **Checks** and **Data** tabs of the proposed change. This is how you preview what the Generator will create before merging.

When `false`, the Generator is skipped entirely in proposed changes. Use this for Generators that are triggered by events (checksum triggers) to avoid double execution — once by CI and once by the trigger.

### `execute_after_merge` (default: `true`)

When `true`, the Generator runs after a merge to apply changes to the target branch.

When `false`, the Generator is skipped after merges. This is rare — mainly useful for Generators that exist solely for preview or validation purposes in proposed changes.

### Common combinations

| `execute_in_proposed_change` | `execute_after_merge` | Use case |
|---|---|---|
| `true` | `true` | **Default.** Full lifecycle — preview in proposed changes, apply on merge. |
| `false` | `true` | Skip preview, only run on merge. Use for Generators with external side effects. |

> | `true` | `false` | Preview only — validate in proposed changes but do not apply on merge. Rare. |

Reviewer: This section isn't clear to me, when a generator runs in a branch (through a proposed change or not) the data it generates aren't part of the merge? If yes then it goes beyond the preview

> | `false` | `false` | Never runs automatically. Only manual execution via `infrahubctl` or event triggers. Use this for event-driven Generators in a [modular setup](modular-generators). |

Reviewer: ```suggestion
| `false` | `false` | Never runs automatically. Only manual execution (via `infrahubctl` / UI...) or event triggers. Use this for event-driven Generators in a [modular setup](modular-generators). |
```
